students/Dustin_L/lesson03/list_lab.py

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python3
+"""List Lab
+
+This module contains all of the functions for the List lab.
+"""
+
+import copy
+
+PROMPT = ' --> '
+
+
+def series_one(lst):
+    """Perform a set of operations on a list of fruit.
+
+    A deep copy is made of the passed list. A series of modifications are made
+    to this copy and the copy is finally returned.
+
+    Args:
+        lst (list): List to be operated upon.
+
+    Returns:
+        list: Modified list.
+    """
+    l = copy.deepcopy(lst)
+    print(l)
+
+    # Prompt user for fruit and add to list
+    usr_in = input('Enter fruit to add to list' + PROMPT)
+    l.append(usr_in)
+    print(l)
+
+    # Prompt user for number and reply with associated fruit
+    usr_in = int(input('Enter a number' + PROMPT))
+    if usr_in <= len(l) and usr_in != 0:
+        print(l[usr_in - 1])
+    else:
+        print('Invalid number!')
+
+    # Add fruit using '+' operator
+    l += ['Grape']
+    print(l)
+
+    # Add fruit using 'insert()' function
+    l.insert(0, 'Mango')
+    print(l)
+
+    # Display all fruit that start with 'P'
+    for fruit in l:
+        if fruit.startswith('P'):
+            print(fruit)
+
+    return l
+
+
+def series_two(lst):
+    """Performs a set of operations on list of fruit.
+
+    A deepy copy is made of the passed list. The last item is removed from this
+    copy. The user is then prompted to selet an item to be removed from list and
+    all specified items are removed. If none are found, the user is continually
+    prompted until the item is found in the list.
+
+    Args:
+        lst (list): List to be operated upon.
+
+    Returns:
+        list: Modified list.
+    """
+    l = copy.deepcopy(lst)
+    print(l)
+
+    # Remove last item
+    l.pop()
+    print(l)
+
+    # Prompt user for fruit and then delete all occurences from list
+    l *= 2
+    while True:
+        usr_in      = input('Enter fruit to remove from list' + PROMPT)
+        removed_lst = [i for i in l if i.lower() != usr_in.lower()]
+
+        if removed_lst != l:
+            l = removed_lst
```

Comment:
You don't necessarily need to reassign `l` here, you could just print and return `removed_list` at this point

Reply:
I see. I'm used to programming in C, where removed_list would be scoped to the While statement and not accessible outside of it (since it is declared in that scope). Is that not the case with Python? What exactly are the scoping rules for Python?
Thanks!

Reply:
Yeah that makes sense. I learned programming and continue to mostly work in Python (and a little JavaScript). The subtle differences between languages are interesting. 

In Python, a function body is the inner most possible scope, so variables defined on the final iteration of a loop remain accessible within the enclosing function body. 

Here are a couple of pretty in-depth articles about scoping in Python that I just found online if you want to get into the nitty-gritty:

http://sebastianraschka.com/Articles/2014_python_scope_and_namespaces.html

https://eli.thegreenplace.net/2015/the-scope-of-index-variables-in-pythons-for-loops/

Reply:
That's great info, thanks!